Add jaguar-document-store.md

fserv opened the pull request:
This PR is attempting to add a new document store for haystack framework. The jaguar document store is a distributed database that can be scaled horizontally easily (instant horizontal scaling with its ZeroMove mechanism). It can store documents, vectors, and blobs data. Also it is capable to detect anomalous documents. It can search similar documents with time decay modulations. The software supports multi-tenant multi-member, single-tenant multi-member, and single-tenant single-member cloud operations models.

bilgeyucel reviewed (changes requested):
Hi @fserv, thanks a ton for developing this integration! 🚀 I've left some comments on the page 👍 

Additionally, I wanted to highlight a couple of points:
1. It seems like the implementation is based on Haystack 2.0, but the examples currently reflect Haystack 1.x. Updating the examples to align with Haystack 2.0 would be great. You can check OpenSearch examples to see the difference between Haystack 1.x and 2.0: https://haystack.deepset.ai/integrations/opensearch-document-store
2. As of Haystack 2.0, document stores now include Retriever components. It would mean that at least one Retriever component should be integrated into the implementation.

To guide you through these changes, I've shared a link to our documentation page explaining the steps to implement a document store for Haystack 2.0: https://docs.haystack.deepset.ai/v2.0/docs/creating-custom-document-stores

If you encounter any challenges or have questions, feel free to reach out—I'm here to help! Alternatively, you can join our Discord community for more interactive discussions: https://discord.gg/haystack 😊

fserv commented:
@bilgeyucel Thanks so much for the suggested changes which are really helpful. We made changes on file jaguar-document-store.md and added JaguarEmbeddingRetriever in the file retriever.py in the https://github.com/fserv/haystack-integrations, directory  /src/jaguar_haystack/document_stores. Other suggested changes are also added in the new commit. Could you please review again? Thanks.

bilgeyucel commented on integrations/jaguar-document-store.md:
````diff
@@ -0,0 +1,124 @@
+Once you saved documents in your `JaguarDocumentStore`, you can use a Haystack pipeline for query.
+
+```python
+from haystack.nodes import AnswerParser, EmbeddingRetriever, PromptNode, PromptTemplate
+
+retriever = EmbeddingRetriever(document_store = document_store,
+                               embedding_model="sentence-transformers/multi-qa-mpnet-base-dot-v1")
+
+prompt = """"Given the provided Documents, answer the Query. Make your answer detailed and long\n
+             Query: {query}\n
+             Documents: {join(documents)}
+             Answer: 
+         """
+prompt_template = PromptTemplate(prompt = prompt, output_parser=AnswerParser())
+prompt_node = PromptNode(model_name_or_path = "gpt-4",
+                         api_key = "YOUR_OPENAI_KEY",
+                         default_prompt_template = prompt_template)
+
+query_pipeline = Pipeline()
+query_pipeline.add_node(component=retriever, name="Retriever", inputs=["Query"])
+query_pipeline.add_node(component=prompt_node, name="PromptNode", inputs=["Retriever"])
+
+query_pipeline.run(query = "Where is the Bermuda Triangle? ", params={"Retriever" : {"top_k": 5}})
````
The showcased code is for Haystack 1.x, the Haystack 2.0 pipeline with JaguarEmbeddingRetriever should look like this 👇 
```suggestion
from haystack import Pipeline
from haystack.components.embedders import SentenceTransformersTextEmbedder
from jaguar_haystack import JaguarEmbeddingRetriever

query_pipeline = Pipeline()
query_pipeline.add_component("text_embedder", SentenceTransformersTextEmbedder())
query_pipeline.add_component("retriever", JaguarEmbeddingRetriever(document_store=document_store))
query_pipeline.connect("text_embedder.embedding", "retriever.query_embedding")

query = "Where is the Bermuda Triangle?"
result = query_pipeline.run({"text_embedder":{"text": query}})
```

bilgeyucel commented on integrations/jaguar-document-store.md:
````diff
@@ -0,0 +1,124 @@
+#### Write Documents
+
+```python
+from haystack.dataclasses import Document
````
This code snippet raises this error
```
E0005 error run(textcol vdb.haystack_test_store.v)
Traceback (most recent call last):
  File "/Users/bilgeyucel/Desktop/deepset/haystack-cookbook/test.py", line 35, in <module>
    document_store.write_documents(documents=docs)
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/jaguar_haystack/jaguar.py", line 124, in write_documents
    zid = self.add_text(text, embedding, metadata, **kwargs)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/jaguar_haystack/jaguar.py", line 257, in add_text
    textcol = js["data"]
              ~~^^^^^^^^
KeyError: 'data'
```

bilgeyucel reviewed (commented):
I reviewed the PR @fserv 👍 It seems like the breaking changes introduced with [2.0.0-beta.5](https://github.com/deepset-ai/haystack/releases/tag/v2.0.0-beta.5) affected the JaguarDocumentStore implementation. These are mainly about changing the import paths
When I downgraded the haystack-ai version to `haystack-ai==2.0.0b4` to eliminate import path issues, there have been some other errors. I left the error messages with some possible solutions as comments. 
I hope they're helpful. Let me know if you have questions :)

fserv commented:
@bilgeyucel Thanks for all the pointers and suggestions! All needed changes are made and pushed. The default  "all-mpnet-base-v2" embedding model has dimension of 768, which may have caused the js['data'] error. Also please make sure the jaguar server and its http gateway server are up-running after the "docker pull jaguardb/jaguardb_with_http;  docker run -d -p 8888:8888 -p 8080:8080 --name jaguardb_with_http  jaguardb/jaguardb_with_http" commands (may require sudo on your system).

fserv commented:
Hi @bilgeyucel Sorry. The package is now updated. You can try "pip install -U  jaguar-haystack" to get the latest package. Thanks!

bilgeyucel commented:
Hi @fserv, I can't seem to `write_documents()` into `JaguarDocumentStore` even with the new version. Here's the error:
```shell
Traceback (most recent call last):
  File "/Users/bilgeyucel/Documents/side-projects/jaguar-haystack/test.py", line 28, in <module>
    document_store.write_documents(documents=result["documents"]) # should write the documents
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/jaguar_haystack/jaguar.py", line 124, in write_documents
    zid = self.add_text(text, embedding, metadata, **kwargs)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/jaguar_haystack/jaguar.py", line 260, in add_text
    textcol = js["data"]
              ~~^^^^^^^^
KeyError: 'data'
```
Maybe there's something wrong how I run the docker image. These are the logs that I get in the container. I'm using MacBook Air M2
```shell
2024-01-30 21:17:03 Starting jaguardb in docker container
2024-01-30 21:17:44 Starting fwww/http server in docker container
2024-01-30 21:17:44 Restart netmap_server ...
2024-01-30 21:17:44 netmap_server is not running
2024-01-30 21:17:46 Restart pyai_server ...
2024-01-30 21:17:46 pyai_server is not running
2024-01-30 21:17:46 netmap_server is not running
2024-01-30 21:17:48 Restart lighttpd and fwww ...
2024-01-30 21:17:48 Stopping fwww ...
2024-01-30 21:17:48 pkill is /usr/bin/pkill
2024-01-30 21:17:48 pyai_server is not running
2024-01-30 21:17:49 Name: sentence-transformers
2024-01-30 21:17:49 Version: 2.2.2
2024-01-30 21:17:49 Summary: Multilingual text embeddings
2024-01-30 21:17:49 Home-page: https://github.com/UKPLab/sentence-transformers
2024-01-30 21:17:49 Author: Nils Reimers
2024-01-30 21:17:49 Author-email: info@nils-reimers.de
2024-01-30 21:17:49 License: Apache License 2.0
2024-01-30 21:17:49 Location: /usr/local/lib/python3.10/dist-packages
2024-01-30 21:17:49 Requires: huggingface-hub, nltk, numpy, scikit-learn, scipy, sentencepiece, torch, torchvision, tqdm, transformers
2024-01-30 21:17:49 Required-by: 
2024-01-30 21:17:49 Found sentence-transformers pip package, OK
2024-01-30 21:17:50 /home/jaguar/fwww/conf_dir/lighttpd.conf is found, OK
2024-01-30 21:17:50 /home/jaguar/fwww/bin_dir/lighttpd -f /home/jaguar/fwww/conf_dir/lighttpd.conf
```

fserv commented:
hi @bilgeyucel most likely one of pyai_server, or netmap_server, or fwww server process is not running properly. You can try:

1. docker exec -it jaguardb_with_http /bin/bash
2. ps aux|grep netmap
3. ps aux|grep pyai
4. ps aux|grep lighttp
5. ps aux|grep fwww

If any server process is not up, you can do this:

# cd /home/jaguar/fwww/bin_dir
# ./start_all_servers.sh

and check again with the "ps aux|grep ..." above. There might be package issues, etc.

fserv commented:
hi @bilgeyucel We did some debugging and found out our documentation missed the document_store.create() step. Sorry for this error. The server startup messages are just for reporting purposes which can be ignored. We checked Mac system docker container and saw all server processes were started fine.  The create() step is added in the latest commit. Please add the create() step in your script too. Thanks!

bilgeyucel reviewed (commented):
Hi @fserv, with your last additions, the `js["data"]` error is gone 🎉 I think we're getting closer to merging this! 

I'll have a look again once the Retriever issue is solved :)

fserv commented on integrations/jaguar-document-store.md:
````diff
@@ -0,0 +1,122 @@
+```python
+from haystack import Pipeline
+from haystack.components.embedders import SentenceTransformersTextEmbedder
+from jaguar_haystack import JaguarEmbeddingRetriever
````
@bilgeyucel Good catch, this is updated.

fserv commented on integrations/jaguar-document-store.md:
```diff
@@ -0,0 +1,122 @@
+
+query_pipeline = Pipeline()
+query_pipeline.add_component("text_embedder", SentenceTransformersTextEmbedder())
+query_pipeline.add_component("retriever", JaguarEmbeddingRetriever(document_store=document_store))
```
@bilgeyucel It is an typo and actually a Dict. It has been corrected and the pushed, Also the package "jaguar-haystack" is updated. You can "pip install -U jaguar-haystack" to have it updated on your platform. Thanks!

bilgeyucel replied:
The same line now throws this error:
```shell
Traceback (most recent call last):
  File "/Users/bilgeyucel/Documents/side-projects/jaguar-haystack/test.py", line 63, in <module>
    query_pipeline.add_component("retriever", JaguarEmbeddingRetriever(document_store=document_store))
                                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: JaguarEmbeddingRetriever.__init__() missing 6 required positional arguments: 'pod', 'store', 'vector_index', 'vector_type', 'vector_dimension', and 'url'
```
'pod', 'store', 'vector_index', 'vector_type', 'vector_dimension', and 'url' infos are already provided at initialization of `document_store`. Do I need to pass them again? 

Also, could you please make sure that the given codes on this page work without any errors before we move forward?

fserv replied:
@bilgeyucel sorry for that. There was differences between v1 and v2 and we had problems running the example. Now the jaguar-haystack is updated please update on your side. Now we still have issue:

from haystack import default_to_dict, default_from_dict
ImportError: cannot import name 'default_to_dict' from 'haystack' 

Are you aware of what might be causing this? Anyway, you could try one more time on your side.

bilgeyucel commented:
Any updates @fserv ?

bilgeyucel commented:
Closing the PR
